scipy/optimize/shgo_m/sobol_seq.py

```diff
@@ -0,0 +1,371 @@
+
+          Output, real R(M,N), the points.
+        """
+        r = np.full((n, dim_num), np.nan)
+        for j in range(n):
+            seed = j + skip
+            r[j, 0:dim_num], next_seed = self.i4_sobol(dim_num, seed)
+
+        return r
+
+    def i4_bit_hi1(self, n):
+        """
+        i4_bit_hi1 returns the position of the high 1 bit base 2 in an integer.
+
+        Example:
+          +------+-------------+-----
+          |    N |      Binary | BIT
+          +------|-------------+-----
+          |    0 |           0 |   0
+          |    1 |           1 |   1
+          |    2 |          10 |   2
+          |    3 |          11 |   2
+          |    4 |         100 |   3
+          |    5 |         101 |   3
+          |    6 |         110 |   3
+          |    7 |         111 |   3
+          |    8 |        1000 |   4
+          |    9 |        1001 |   4
+          |   10 |        1010 |   4
+          |   11 |        1011 |   4
+          |   12 |        1100 |   4
+          |   13 |        1101 |   4
+          |   14 |        1110 |   4
+          |   15 |        1111 |   4
+          |   16 |       10000 |   5
+          |   17 |       10001 |   5
+          | 1023 |  1111111111 |  10
+          | 1024 | 10000000000 |  11
+          | 1025 | 10000000001 |  11
+
+        Parameters:
+          Input, integer N, the integer to be measured.
+          N should be nonnegative.  If N is nonpositive,
+          the value will always be 0.
+
+          Output, integer BIT, the number of bits base 2.
+        """
+        i = np.floor(n)
+        bit = 0
+        while i > 0:
+            bit += 1
+            i //= 2
+        return bit
+
+    def i4_bit_lo0(self, n):
+        """
+        I4_BIT_LO0 returns the position of the low 0 bit base 2 in an integer.
+
+        Example:
+          +------+------------+----
+          |    N |     Binary | BIT
+          +------+------------+----
+          |    0 |          0 |   1
+          |    1 |          1 |   2
+          |    2 |         10 |   1
+          |    3 |         11 |   3
+          |    4 |        100 |   1
+          |    5 |        101 |   2
+          |    6 |        110 |   1
+          |    7 |        111 |   4
+          |    8 |       1000 |   1
+          |    9 |       1001 |   2
+          |   10 |       1010 |   1
+          |   11 |       1011 |   3
+          |   12 |       1100 |   1
+          |   13 |       1101 |   2
+          |   14 |       1110 |   1
+          |   15 |       1111 |   5
+          |   16 |      10000 |   1
+          |   17 |      10001 |   2
+          | 1023 | 1111111111 |   1
+          | 1024 | 0000000000 |   1
+          | 1025 | 0000000001 |   1
+
+        Parameters:
+          Input, integer N, the integer to be measured.
+          N should be nonnegative.
+
+          Output, integer BIT, the position of the low 1 bit.
+        """
+        bit = 1
```

Comment:
These functions can be much simplified and maintaining them becomes much easier.
```python
def i4_bit_lo0(n):
     # Find the first 0 appearance, if no 0 found return -1
    ind = np.binary_repr(x)[::-1].find('0')

     # If all 1s then return the next bit index
    if ind == -1:
        ind = len(n) + 1
    # Add one to match the 1indexing in the table above
    else:
        ind += 1
    return ind
```